src/account/sync/mod.rs

```diff
@@ -539,14 +539,102 @@ impl SyncedAccount {
         Ok(message)
     }
 
-    /// Retry messages.
-    pub fn retry(&self, message_id: &MessageId) -> crate::Result<Message> {
-        let account: Account =
-            crate::storage::get_account(&self.storage_path, self.account_id.clone().into())?;
-        let message = account
-            .get_message(message_id)
-            .ok_or_else(|| anyhow::anyhow!("transaction with the given id not found"));
-        unimplemented!()
+    /// Retry message.
+    pub async fn retry(&self, message_id: &MessageId) -> crate::Result<Message> {
+        repost_message(
+            self.account_id.clone().into(),
+            &self.storage_path,
+            message_id,
+            RepostAction::Retry,
+        )
+        .await
+    }
+
+    /// Promote message.
+    pub async fn promote(&self, message_id: &MessageId) -> crate::Result<Message> {
+        repost_message(
+            self.account_id.clone().into(),
+            &self.storage_path,
+            message_id,
+            RepostAction::Promote,
+        )
+        .await
+    }
+
+    /// Reattach message.
+    pub async fn reattach(&self, message_id: &MessageId) -> crate::Result<Message> {
+        repost_message(
+            self.account_id.clone().into(),
+            &self.storage_path,
+            message_id,
+            RepostAction::Reattach,
+        )
+        .await
+    }
+}
+
+enum RepostAction {
+    Retry,
+    Reattach,
+    Promote,
+}
+
+async fn repost_message(
+    account_id: AccountIdentifier,
+    storage_path: &PathBuf,
+    message_id: &MessageId,
+    action: RepostAction,
+) -> crate::Result<Message> {
+    let mut account: Account = crate::storage::get_account(&storage_path, account_id)?;
+    match account.get_message(message_id) {
```

Comment:
@lucasfernog At this point, we also need to get the reattachments for this message from the database. We need to check metadata for all messages (original + reattachments). It's possible that one of them got already confirmed and we are looking for metadata for the unconfirmed one.

Reply:
I'm using the `list_messages` algorithm [here](https://github.com/iotaledger/wallet.rs/blob/63ebcb2d9b68e283ca945cc5210be06088ff6802/src/account/sync/mod.rs#L593) to get the actual message we care about and if it's confirmed, we error out